### ISSUE TITLE
Add support for read only transactions

### ISSUE THREAD

**Issue by synedu** [neotypes/neotypes#178]
The neo4j java driver provides a method on AsyncSession to allow a user to specify that a transaction should not mutate data on the neo4j server: [ReadTransactionAsync](https://neo4j.com/docs/driver-manual/current/session-api/asynchronous/). This allows optimisation in clustered environments - such as routing read-only queries to read replicas.

It should be noted that introspection does not occur on the neo4j driver side, and it does not check to see that a query won't mutate data before routing the request, but the read replica will reject it.

**Comment by hugofirth** [comment]
I'd like to volunteer to take this on :)

**Comment by BalmungSan** [comment]
Hi, @hugofirth welcome and thank you very much for the interest in the project :)

There is an already effort in solving this one at #180 it is almost done; we are just waiting for #207 so that can unblock #164 and finally merge master in #180 to solve conflicts and we will be done.

Nevertheless to say that the implementation in #180 is a bit hacky since the official **neo4j** **Java** driver doesn't provide a way to just create a read-only transaction but rather a `readOnlyTransact` which lead to a somewhat tricky implementation.
May I ask if it would be possible to expose a method to create a read-only transaction directly in the **neo4j** **Java** driver, or even better to pass the desired `AccessMode` to `beginTransactionAsync`? And f so, if you could _(and obviously, want to)_ do that work there so we can pick up that here and simplify the implementation of #180  that would be amazing!

BTW, feel free to drop in [**gitter**](https://gitter.im/neotypes-neotypes/Lobby) to chat about this or any other change :)

**Comment by hugofirth** [comment]
I don't work on the drivers team at Neo, but I can pass along the request for sure.

**Comment by BalmungSan** [comment]
@bengraygh due to the work being done in #221 **Sessions** no longer exists and you can specify that the transaction has to be read-only through `withAccessMode(AcessMode.READ)` on **TransactionConfig**.

That should be enough for your use case, but:
+ Implies extra boilerplate from the user side.
+ Isn't automatically used by operations like `single` or `list`.

Do you still think it would be worth to add `readOnlyTransact`? or maybe even `readOnlyTransaction`?

**Comment by bengraygh** [comment]
This is completely fine for my case, and I think for most people. However, I think there is an advantage to specifying read or write transaction at the session level for users in clustered environments. Sessions can contain multiple transactions, some of which are read, and some of which are write, and there is retry logic at the session level. If the gatekeeping is at the Session level, it provides a little more declarative power: none of the transactions I'm giving you can change the database, and all of them should be rejected if they try.

Even if there is utility to it, this is an enhancement IMO and not a high priority item. There's nothing stopping us implementing it down the track.

**Comment by BalmungSan** [comment]
Yeah and since **Session** won't be available for users for a while I believe we are ok for now.

**Comment by BalmungSan** [comment]
_"Fixed"_ as part of #221

It may be good to open a new issue to discuss if common read operations should be read-only. Or if it would be good to add `readOnlyTransact` _(and maybe `readOnlyTransaction`)_.
